Mathlib/Combinatorics/Flow/Basic.lean

```diff
@@ -0,0 +1,232 @@
+  nonneg' : ∀ u v, 0 ≤ toFun u v
+  capacity' : ∀ u v, toFun u v ≤ N.cap u v
+
+namespace PseudoFlow
+
+variable {N : Network V R}
+
+instance instFunLike : FunLike N.PseudoFlow V (V → R) where
+  coe := PseudoFlow.toFun
+  coe_injective' f g h := by cases f; cases g; congr
+
+variable (f : N.PseudoFlow)
+
+lemma nonneg : ∀ u v, 0 ≤ f u v := f.nonneg'
+
+lemma capacity : ∀ u v, f u v ≤ N.cap u v := f.capacity'
+
+instance : Zero N.PseudoFlow where
+  zero.toFun _ _ := 0
+  zero.nonneg' _ _ := le_rfl
+  zero.capacity' := N.nonneg
+
+end PseudoFlow
+end PseudoFlowDefinition
+
+variable [Fintype V]
+
+namespace PseudoFlow
+
+variable [Preorder R] [AddCommMonoid R] {N : Network V R} (f : N.PseudoFlow) (v : V)
+
+/-- The incoming flow of a vertex. --/
+def incoming := ∑ u, f u v
+/-- The outgoing flow of a vertex. --/
+def outgoing := ∑ w, f v w
+/-- The excess flow of a vertex (the incoming flow minus the outgoing flow). --/
+def excess [Sub R] := f.incoming v - f.outgoing v
+
+end PseudoFlow
+
+section PreFlowDefinition
+
+variable [Preorder R] [AddCommMonoid R] (N : Network V R) (s t : V)
+
+/-- A `PreFlow` from `s` to `t` is a `PseudoFlow` such that all vertices except `s` and `t` have
+non-negative excess, that is, the outgoing flow is at most as much as the incoming flow. -/
+@[ext]
+structure PreFlow extends N.PseudoFlow where
+  excess_nonneg : ∀ v, v ≠ s ∧ v ≠ t → toPseudoFlow.outgoing v ≤ toPseudoFlow.incoming v
+
+namespace PreFlow
+
+variable {N s t}
+
+instance : Zero (N.PreFlow s t) where
+  zero := { (0 : N.PseudoFlow) with excess_nonneg := fun _ _ ↦ le_rfl }
+
+/-- The value of a flow is the amount of flow units that leave the source vertex and arrive at the
+sink vertex. In particular, this is equal to the excess of `t`. -/
+def value [Sub R] (f : N.PreFlow s t) := f.excess t
+
+end PreFlow
+end PreFlowDefinition
+
+section FlowDefinition
+
+variable [Preorder R] [AddCommMonoid R] (N : Network V R) (s t : V)
+
+/-- A `Flow` from `s` to `t` is a `PreFlow s t`, but instead of requiring non-negative excess, it
+requires that the excess is 0 (also called _flow conservation_). -/
+@[ext]
+structure Flow extends N.PreFlow s t where
+  conservation : ∀ v, v ≠ s ∧ v ≠ t → toPseudoFlow.outgoing v = toPseudoFlow.incoming v
+  excess_nonneg v hv := le_of_eq <| conservation v hv
+
+namespace Flow
+
+instance : Zero (N.Flow s t) where
+  zero := { (0 : N.PreFlow s t) with conservation := fun _ _ ↦ rfl }
+
+end Flow
+end FlowDefinition
+
+section SummingLemmas
+variable [Preorder R] [AddCommGroup R] {N : Network V R}
+
+namespace PseudoFlow
+
+variable (f : N.PseudoFlow)
+
+lemma sum_outgoing_eq_sum_incoming : ∑ v, f.outgoing v = ∑ v, f.incoming v := by
+  unfold outgoing incoming
+  rw[Finset.sum_comm]
+
+@[simp]
+lemma sum_excess_zero : ∑ v, f.excess v = 0 := by simp[excess, sum_outgoing_eq_sum_incoming]
+
+end PseudoFlow
+
+namespace Flow
+
+variable {s t : V} (f : N.Flow s t)
+
+lemma excess_s_zero_of_eq (hst : s = t) : f.excess s = 0 := by
```

Comment:
```suggestion
lemma excess_source_zero_of_eq (hst : s = t) : f.excess s = 0 := by
```
or maybe just `excess_zero_of_eq`. In general it's better to avoid referencing variable names in your lemma names - remember that the person using the lemma might not have used the same name as you did! Instead, giving a semantic meaning can be better

Reply:
fair enough, `s` and `t` where ubiquitous for source and sink for me so far, but your suggestion is better. Should we go with `excess_{source, sink}_zero_of_eq` or with `excess_zero_of_eq` and `excess_zero_of_eq'`? I think I slightly prefer the first option